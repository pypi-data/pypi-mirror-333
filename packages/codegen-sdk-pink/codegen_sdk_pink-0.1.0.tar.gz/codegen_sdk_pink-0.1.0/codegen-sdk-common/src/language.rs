use std::{hash::Hash, num::NonZeroU16, path::PathBuf, sync::Arc};

use convert_case::{Case, Casing};
use mockall::automock;
use proc_macro2::Span;
use tree_sitter::Parser;

use crate::{
    errors::ParseError,
    naming::normalize_type_name,
    parser::{Node, parse_node_types},
};
#[derive(Debug, Eq, PartialEq)]
pub struct Language {
    name: &'static str,
    pub struct_name: &'static str,
    pub node_types: &'static str,
    pub file_extensions: &'static [&'static str],
    tree_sitter_language: tree_sitter::Language,
    pub tag_query: &'static str,
    nodes: Vec<Arc<Node>>,
}
impl Hash for Language {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.name.hash(state);
    }
}

#[automock]
impl Language {
    pub fn new(
        name: &'static str,
        struct_name: &'static str,
        node_types: &'static str,
        file_extensions: &'static [&'static str],
        tree_sitter_language: tree_sitter::Language,
        tag_query: &'static str,
    ) -> anyhow::Result<Self> {
        let nodes = parse_node_types(node_types)?
            .into_iter()
            .map(|node| Arc::new(node))
            .collect();
        Ok(Self {
            name,
            struct_name,
            node_types,
            file_extensions,
            tree_sitter_language,
            tag_query,
            nodes,
        })
    }
    pub fn parse_tree_sitter(&self, content: &str) -> Result<tree_sitter::Tree, ParseError> {
        let mut parser = Parser::new();
        parser.set_language(&self.tree_sitter_language)?;
        parser.parse(content, None).ok_or(ParseError::Miscelaneous)
    }
    pub fn nodes(&self) -> &Vec<Arc<Node>> {
        &self.nodes
    }
    pub fn root_node(&self) -> String {
        self.nodes()
            .iter()
            .find(|node| node.root)
            .unwrap_or_else(|| panic!("No root node found for language: {}", self.name))
            .type_name
            .to_case(Case::Pascal)
    }
    pub fn kind_id(&self, name: &str, named: bool) -> u16 {
        self.tree_sitter_language.id_for_node_kind(name, named)
    }
    pub fn kind_name(&self, id: u16) -> Option<&'static str> {
        self.tree_sitter_language.node_kind_for_id(id)
    }
    pub fn field_id(&self, name: &str) -> Option<NonZeroU16> {
        self.tree_sitter_language.field_id_for_name(name)
    }
    pub fn field_name(&self, id: u16) -> Option<&'static str> {
        self.tree_sitter_language.field_name_for_id(id)
    }
    pub fn name(&self) -> &'static str {
        self.name
    }
    pub fn struct_name(&self) -> &'static str {
        self.struct_name
    }
    pub fn file_struct_name(&self) -> syn::Ident {
        syn::Ident::new(&format!("{}File", self.struct_name()), Span::call_site())
    }
    pub fn package_name(&self) -> String {
        format!("codegen_sdk_{}", self.name())
    }
    pub fn node_for_struct_name(&self, struct_name: &str) -> Option<Arc<Node>> {
        self.nodes
            .iter()
            .find(|node| normalize_type_name(&node.type_name, node.named) == struct_name)
            .cloned()
    }
    pub fn should_parse(&self, file_path: &PathBuf) -> Result<bool, ParseError> {
        Ok(self.file_extensions.contains(
            &file_path
                .extension()
                .ok_or(ParseError::Miscelaneous)?
                .to_str()
                .ok_or(ParseError::Miscelaneous)?,
        ))
    }
}
#[cfg(feature = "go")]
pub mod go;
#[cfg(feature = "java")]
pub mod java;
#[cfg(feature = "typescript")]
pub mod javascript;
#[cfg(feature = "json")]
pub mod json;
#[cfg(feature = "typescript")]
pub mod jsx;
#[cfg(feature = "markdown")]
pub mod markdown;
#[cfg(feature = "python")]
pub mod python;
#[cfg(feature = "ruby")]
pub mod ruby;
#[cfg(feature = "rust")]
pub mod rust;
#[cfg(feature = "toml")]
pub mod toml;
#[cfg(feature = "ts_query")]
pub mod ts_query;
#[cfg(feature = "typescript")]
pub mod tsx;
#[cfg(feature = "typescript")]
pub mod typescript;
#[cfg(feature = "yaml")]
pub mod yaml;
lazy_static! {
    pub static ref LANGUAGES: Vec<&'static Language> = vec![
        #[cfg(feature = "python")]
        &python::Python,
        #[cfg(feature = "typescript")]
        &typescript::Typescript,
        #[cfg(feature = "typescript")]
        &tsx::TSX,
        #[cfg(feature = "typescript")]
        &jsx::JSX,
        #[cfg(feature = "typescript")]
        &javascript::Javascript,
        #[cfg(feature = "rust")]
        &rust::Rust,
        #[cfg(feature = "go")]
        &go::Go,
        #[cfg(feature = "ruby")]
        &ruby::Ruby,
        #[cfg(feature = "yaml")]
        &yaml::Yaml,
        #[cfg(feature = "toml")]
        &toml::TOML,
        #[cfg(feature = "markdown")]
        &markdown::Markdown,
        #[cfg(feature = "json")]
        &json::JSON,
        #[cfg(feature = "java")]
        &java::Java,
        #[cfg(feature = "ts_query")]
        &ts_query::Query,
    ];
}
