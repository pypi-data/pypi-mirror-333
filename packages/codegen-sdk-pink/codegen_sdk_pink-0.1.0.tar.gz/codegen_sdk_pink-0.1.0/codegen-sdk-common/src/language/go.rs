use super::Language;
lazy_static! {
    pub static ref Go: Language = Language::new(
        "go",
        "Go",
        tree_sitter_go::NODE_TYPES,
        &["go"],
        tree_sitter_go::LANGUAGE.into(),
        tree_sitter_go::TAGS_QUERY,
    )
    .unwrap();
}
