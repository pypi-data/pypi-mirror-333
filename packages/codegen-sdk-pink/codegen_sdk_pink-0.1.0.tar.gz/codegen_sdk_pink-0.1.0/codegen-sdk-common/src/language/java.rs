use super::Language;
lazy_static! {
    pub static ref Java: Language = Language::new(
        "java",
        "Java",
        tree_sitter_java::NODE_TYPES,
        &["java"],
        tree_sitter_java::LANGUAGE.into(),
        tree_sitter_java::TAGS_QUERY,
    )
    .unwrap();
}
