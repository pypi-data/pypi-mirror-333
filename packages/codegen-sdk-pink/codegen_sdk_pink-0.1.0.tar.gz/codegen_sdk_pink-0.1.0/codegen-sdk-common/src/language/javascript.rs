use super::Language;

lazy_static! {
    pub static ref Javascript: Language = Language::new(
        "javascript",
        "Javascript",
        tree_sitter_javascript::NODE_TYPES,
        &["js"],
        tree_sitter_javascript::LANGUAGE.into(),
        tree_sitter_javascript::TAGS_QUERY,
    )
    .unwrap();
}
