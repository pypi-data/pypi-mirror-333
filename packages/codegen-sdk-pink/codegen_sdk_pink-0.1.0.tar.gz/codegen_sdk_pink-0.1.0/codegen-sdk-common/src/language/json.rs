use super::Language;
lazy_static! {
    pub static ref JSON: Language = Language::new(
        "json",
        "JSON",
        tree_sitter_json::NODE_TYPES,
        &["json"],
        tree_sitter_json::LANGUAGE.into(),
        "",
    )
    .unwrap();
}
