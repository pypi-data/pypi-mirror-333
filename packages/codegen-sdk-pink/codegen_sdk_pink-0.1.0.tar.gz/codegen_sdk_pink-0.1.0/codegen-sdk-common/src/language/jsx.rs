use super::Language;
lazy_static! {
    pub static ref JSX: Language = Language::new(
        "jsx",
        "JSX",
        tree_sitter_typescript::TSX_NODE_TYPES,
        &["jsx"],
        tree_sitter_typescript::LANGUAGE_TSX.into(),
        tree_sitter_typescript::TAGS_QUERY,
    )
    .unwrap();
}
