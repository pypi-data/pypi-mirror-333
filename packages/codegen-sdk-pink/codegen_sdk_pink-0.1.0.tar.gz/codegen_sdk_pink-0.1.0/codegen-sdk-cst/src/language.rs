use std::{hash::Hash, path::PathBuf, sync::Arc};

use bytes::Bytes;
use codegen_sdk_common::{
    ParseError, Tree,
    language::Language,
    traits::{CSTNode, FromNode},
    tree::TreeNode,
};

pub trait CSTLanguage {
    type Types<'db>: TreeNode;
    type Program<'db1>: CSTNode<'db1> + FromNode<'db1, Self::Types<'db1>> + Send;
    fn language() -> &'static Language;
    fn parse<'db>(
        db: &'db dyn salsa::Database,
        content: String,
    ) -> Option<(
        &'db Self::Program<'db>,
        &'db Tree<Self::Types<'db>>,
        indextree::NodeId,
    )>;
    fn parse_file_from_cache<'db>(
        db: &'db dyn salsa::Database,
        file_path: &PathBuf,
        #[cfg(feature = "serialization")] cache: &'db codegen_sdk_common::serialize::Cache,
    ) -> Result<
        Option<(
            &'db Self::Program<'db>,
            &'db Tree<Self::Types<'db>>,
            indextree::NodeId,
        )>,
        ParseError,
    > {
        #[cfg(feature = "serialization")]
        {
            let serialized_path = cache.get_path(file_path);
            if serialized_path.exists() {
                let parsed = cache.read_entry::<Self::Program<'db>>(&serialized_path)?;
                return Ok(Some(parsed));
            }
        }
        Ok(None)
    }
    fn parse_file<'db>(
        db: &'db dyn salsa::Database,
        file_path: &PathBuf,
        #[cfg(feature = "serialization")] cache: &'db codegen_sdk_common::serialize::Cache,
    ) -> Result<
        Option<(
            &'db Self::Program<'db>,
            &'db Tree<Self::Types<'db>>,
            indextree::NodeId,
        )>,
        ParseError,
    > {
        if let Some(parsed) = Self::parse_file_from_cache(
            db,
            file_path,
            #[cfg(feature = "serialization")]
            cache,
        )? {
            return Ok(Some(parsed));
        }
        let content = std::fs::read_to_string(file_path)?;
        if let Some(parsed) = Self::parse(db, content) {
            return Ok(Some(parsed));
        }
        Err(ParseError::SyntaxError)
    }

    fn should_parse(file_path: &PathBuf) -> Result<bool, ParseError> {
        Self::language().should_parse(file_path)
    }
}
