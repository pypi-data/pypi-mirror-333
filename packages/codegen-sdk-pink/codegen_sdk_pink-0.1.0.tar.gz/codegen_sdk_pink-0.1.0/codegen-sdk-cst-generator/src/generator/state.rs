use std::{
    collections::{BTreeMap, BTreeSet, HashMap, HashSet, VecDeque},
    sync::Arc,
};

#[double]
use codegen_sdk_common::language::Language;
use codegen_sdk_common::{naming::normalize_type_name, parser::TypeDefinition};
use mockall_double::double;
use proc_macro2::TokenStream;
use quote::{format_ident, quote};
use syn::Ident;

use super::node::Node;
use crate::{
    Config,
    generator::{
        constants::TYPE_NAME,
        utils::{get_comment_type, get_from_enum_to_ref, get_from_node, get_from_type},
    },
};
#[derive(Debug)]
pub struct State<'a> {
    pub subenums: BTreeSet<String>,
    config: Config,
    nodes: BTreeMap<String, Node<'a>>,
}
impl<'a> State<'a> {
    pub fn new(language: &'a Language, config: Config) -> Self {
        let mut nodes = BTreeMap::new();
        let mut subenums = BTreeSet::new();
        let raw_nodes = language.nodes();
        for raw_node in raw_nodes {
            if raw_node.subtypes.is_empty() {
                let node = Node::new(raw_node.clone(), language, config.clone());
                nodes.insert(node.normalize_name(), node);
            } else {
                subenums.insert(raw_node.type_name.clone());
            }
        }
        let mut ret = Self {
            nodes,
            subenums,
            config,
        };
        let mut subenums = VecDeque::new();
        for raw_node in raw_nodes.iter().filter(|n| !n.subtypes.is_empty()) {
            subenums.push_back(raw_node.clone());
        }
        while let Some(raw_node) = subenums.pop_front() {
            if raw_node
                .subtypes
                .iter()
                .any(|s| subenums.iter().any(|n| n.type_name == s.type_name))
            {
                subenums.push_back(raw_node);
            } else {
                // Add subtypes to the state
                ret.add_subenum(&raw_node.type_name, &raw_node.subtypes.iter().collect());
            }
        }

        log::info!("Adding child subenums");
        ret.add_child_subenums();
        log::info!("Adding missing fields");
        ret.add_missing_fields(language);
        log::info!("Adding field subenums");
        ret.add_field_subenums();
        ret
    }
    fn add_missing_fields(&mut self, language: &'a Language) {
        let mut to_insert = Vec::new();
        for node in self.nodes.values() {
            for field in &node.fields {
                for type_def in field.types() {
                    let name = type_def.normalize();
                    if !self.nodes.contains_key(&name)
                        && !self.subenums.contains(&type_def.type_name)
                    {
                        let node = codegen_sdk_common::parser::Node {
                            type_name: type_def.type_name.clone(),
                            subtypes: vec![],
                            named: type_def.named,
                            root: false,
                            fields: None,
                            children: None,
                        };
                        to_insert.push((name, node));
                    }
                }
            }
        }
        to_insert.dedup_by_key(|(name, _)| name.clone());
        for (name, node) in to_insert {
            self.nodes.insert(
                name,
                Node::new(Arc::new(node), language, self.config.clone()),
            );
        }
    }
    fn add_child_subenums(&mut self) {
        let keys = self.nodes.keys().cloned().collect::<Vec<_>>();
        for name in keys.into_iter() {
            let normalized_name = normalize_type_name(&name, true);
            let node = self.nodes.get(&normalized_name).unwrap();
            let mut children_types = node.get_children_names();
            if children_types.len() > 1 {
                children_types.sort();
                children_types.dedup();
                self.add_subenum(
                    &node.children_struct_name(),
                    &children_types.iter().collect(),
                );
            }
        }
    }
    fn add_field_subenums(&mut self) {
        let mut to_add: Vec<(String, Vec<TypeDefinition>)> = Vec::new();
        for node in self.nodes.values() {
            for field in &node.fields {
                log::debug!("Adding field subenum: {}", field.normalized_name());
                if field.types().len() > 1 {
                    to_add.push((
                        field.type_name(),
                        field.types().into_iter().cloned().collect(),
                    ));
                }
            }
        }
        for (name, types) in to_add.into_iter() {
            self.add_subenum(&name, &types.iter().collect());
        }
    }
    fn add_subenum(&mut self, name: &str, nodes: &Vec<&TypeDefinition>) {
        self.subenums.insert(name.to_string());
        let mut nodes = nodes.clone();
        let comment = get_comment_type();
        if self.nodes.contains_key(&comment.type_name) {
            nodes.push(&comment);
        }
        for node in nodes {
            let normalized_name = normalize_type_name(&node.type_name, node.named);
            if !self.subenums.contains(&node.type_name) {
                log::debug!("Adding subenum: {} to {}", name, normalized_name);
                if let Some(node) = self.nodes.get_mut(&normalized_name) {
                    node.add_subenum(name.to_string());
                }
            } else {
                let variants = self.get_variants(&node.type_name, true);
                self.add_subenum(name, &variants.iter().collect());
            }
        }
    }
    pub fn get_variants(&self, subenum: &str, include_comment: bool) -> Vec<TypeDefinition> {
        let mut variants = Vec::new();
        for node in self.nodes.values() {
            log::debug!("Checking subenum: {} for {}", subenum, node.kind());
            if node.subenums.contains(&subenum.to_string()) {
                log::debug!("Found variant: {} for {}", node.kind(), subenum);
                variants.push(node.type_definition());
            }
        }
        if include_comment {
            let comment = get_comment_type();
            if !variants.iter().any(|v| v.type_name == comment.type_name) {
                variants.push(comment);
            }
        }
        variants
    }
    fn get_variant_map(&self, enum_name: &str) -> BTreeMap<u16, TokenStream> {
        let mut variant_map = BTreeMap::new();
        for node in self.nodes.values() {
            let variant_name = format_ident!("{}", node.normalize_name());
            if node.subenums.contains(&enum_name.to_string()) {
                log::debug!("Adding variant: {} for {}", node.kind(), enum_name);
                variant_map.insert(
                    node.kind_id(),
                    quote! {
                        {
                            let (node, ids) = #variant_name::from_node(context, node)?;
                            Ok((Self::#variant_name(node), ids))
                        }
                    },
                );
            }
        }
        variant_map
    }
    // Implement the TSNode => CSTNode conversion trait on a given subenum
    fn get_from_node(&self, enum_name: &str) -> TokenStream {
        let variant_map = self.get_variant_map(enum_name);
        get_from_node(enum_name, true, &variant_map)
    }
    // Get the overarching enum for the nodes
    pub fn get_enum(&self, is_ref: bool) -> TokenStream {
        let mut enum_tokens = Vec::new();
        let mut from_tokens = TokenStream::new();
        let mut subenums = Vec::new();
        let mut subenum_name_map = HashMap::new();
        for name in self.subenums.iter() {
            subenum_name_map.insert(name.clone(), normalize_type_name(name, true));
        }
        let enum_name = if is_ref {
            format_ident!("{}Ref", TYPE_NAME)
        } else {
            format_ident!("{}", TYPE_NAME)
        };
        for node in self.nodes.values() {
            enum_tokens.push(node.get_enum_tokens(&subenum_name_map, is_ref));
            from_tokens.extend_one(get_from_type(&node.normalize_name(), &enum_name, is_ref));
        }
        if is_ref {
            let variants = self.get_node_struct_names();
            let ref_convert = get_from_enum_to_ref(&TYPE_NAME, &variants);
            from_tokens.extend_one(ref_convert);
        }
        for subenum in self.subenums.iter() {
            assert!(
                self.get_variants(subenum, true).len() > 0,
                "Subenum {} has no variants",
                subenum
            );
            if is_ref {
                let normalized_subenum = normalize_type_name(&subenum, true);
                let ref_name = format_ident!("{}Ref", normalized_subenum);
                subenums.push(ref_name);
                let variants = self
                    .get_variants(subenum, false)
                    .into_iter()
                    .map(|v| v.ident())
                    .collect();
                let ref_convert = get_from_enum_to_ref(&normalized_subenum, &variants);
                from_tokens.extend_one(ref_convert);
            } else {
                from_tokens.extend_one(self.get_from_node(subenum));
                subenums.push(format_ident!("{}", normalize_type_name(&subenum, true)));
            }
        }
        let subenum_tokens = if !subenums.is_empty() {
            subenums.sort();
            subenums.dedup();
            if self.config.serialize {
                quote! {
                    #[subenum(#(#subenums(derive(Archive, Deserialize, Serialize))),*)]
                }
            } else {
                quote! {
                    #[subenum(#(#subenums),*)]
                }
            }
        } else {
            quote! {}
        };
        let derive_header = if is_ref {
            quote! {
                #[derive(Debug, Eq, PartialEq, Hash, Clone, Ord, PartialOrd)]
            }
        } else {
            quote! {
            #[derive(Debug, Eq, PartialEq,Clone, Hash, salsa::Update, Delegate, Ord, PartialOrd)]
            #[delegate(
                CSTNode<'db1>
            )]
                }
        };
        quote! {
            #subenum_tokens
            #derive_header
        pub enum #enum_name<'db1> {
                #(#enum_tokens),*
            }
            #from_tokens
        }
    }
    pub fn get_structs(&self) -> TokenStream {
        let mut struct_tokens = TokenStream::new();
        let subenums = HashSet::from_iter(self.subenums.iter());
        for node in self.nodes.values() {
            struct_tokens.extend_one(node.get_struct_tokens(&subenums));
        }
        struct_tokens
    }
    pub fn get_node_for_struct_name(&self, name: &str) -> Option<&Node<'a>> {
        self.nodes.get(name)
    }
    pub fn get_node_for_raw_name(&self, name: &str) -> Option<&Node<'a>> {
        for node in self.nodes.values() {
            if node.kind() == name {
                return Some(node);
            }
        }
        None
    }
    pub fn get_subenum_variants(&self, name: &str, include_comment: bool) -> Vec<&Node<'a>> {
        let variants = self.get_variants(name, include_comment);
        let mut nodes = Vec::new();
        for variant in variants {
            if let Some(node) = self.get_node_for_struct_name(&variant.normalize()) {
                nodes.push(node);
            }
        }
        nodes
    }
    pub fn get_node_struct_names(&self) -> Vec<Ident> {
        self.nodes
            .values()
            .map(|node| format_ident!("{}", node.normalize_name()))
            .collect()
    }
    pub fn get_subenum_struct_names(&self) -> Vec<Ident> {
        self.subenums
            .iter()
            .map(|s| format_ident!("{}", normalize_type_name(s, true)))
            .collect()
    }
    pub fn nodes(&self) -> Vec<&Node<'a>> {
        self.nodes.values().collect()
    }
}
#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use super::*;
    use crate::test_util::{get_language, snapshot_tokens};

    #[test_log::test]
    fn test_get_enum() {
        let node = codegen_sdk_common::parser::Node {
            type_name: "test".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        };
        let nodes = vec![node];
        let language = get_language(nodes);
        let state = State::new(&language, Config::default());
        let enum_tokens = state.get_enum(false);
        insta::assert_debug_snapshot!(snapshot_tokens(&enum_tokens));
    }
    #[test_log::test]
    fn test_parse_children() {
        let child = codegen_sdk_common::parser::Node {
            type_name: "child".to_string(),
            subtypes: vec![],
            named: false,
            root: false,
            fields: None,
            children: None,
        };
        let child_two = codegen_sdk_common::parser::Node {
            type_name: "child_two".to_string(),
            subtypes: vec![],
            named: false,
            root: false,
            fields: None,
            children: None,
        };
        let node = codegen_sdk_common::parser::Node {
            type_name: "test".to_string(),
            subtypes: vec![],
            named: false,
            root: false,
            fields: None,
            children: Some(codegen_sdk_common::parser::Children {
                multiple: true,
                required: false,
                types: vec![
                    codegen_sdk_common::parser::TypeDefinition {
                        type_name: "child".to_string(),
                        named: true,
                    },
                    codegen_sdk_common::parser::TypeDefinition {
                        type_name: "child_two".to_string(),
                        named: true,
                    },
                ],
            }),
        };
        let nodes = vec![child, child_two, node];
        let language = get_language(nodes);
        let state = State::new(&language, Config::default());
        let enum_tokens = state.get_enum(false);
        insta::assert_debug_snapshot!(snapshot_tokens(&enum_tokens));
    }
    #[test_log::test]
    fn test_parse_children_subtypes() {
        let definition = codegen_sdk_common::parser::Node {
            type_name: "definition".to_string(),
            subtypes: vec![
                TypeDefinition {
                    type_name: "function".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "class".to_string(),
                    named: true,
                },
            ],
            named: true,
            root: false,
            fields: None,
            children: None,
        };
        let class = codegen_sdk_common::parser::Node {
            type_name: "class".to_string(),
            subtypes: vec![],
            named: false,
            root: false,
            fields: None,
            children: Some(codegen_sdk_common::parser::Children {
                multiple: true,
                required: false,
                types: vec![codegen_sdk_common::parser::TypeDefinition {
                    type_name: "definition".to_string(),
                    named: true,
                }],
            }),
        };
        let function = codegen_sdk_common::parser::Node {
            type_name: "function".to_string(),
            subtypes: vec![],
            named: false,
            root: false,
            fields: None,
            children: None,
        };
        let nodes = vec![definition, class, function];
        let language = get_language(nodes);
        let state = State::new(&language, Config::default());
        let enum_tokens = state.get_enum(false);
        insta::assert_debug_snapshot!(snapshot_tokens(&enum_tokens));
    }
    #[test_log::test]
    fn test_add_field_subenums() {
        let node_a = codegen_sdk_common::parser::Node {
            type_name: "node_a".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        };
        let node_b = codegen_sdk_common::parser::Node {
            type_name: "node_b".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        };
        let field = codegen_sdk_common::parser::FieldDefinition {
            types: vec![
                TypeDefinition {
                    type_name: "node_a".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "node_b".to_string(),
                    named: true,
                },
            ],
            multiple: true,
            required: false,
        };
        let node_c = codegen_sdk_common::parser::Node {
            type_name: "node_c".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: Some(codegen_sdk_common::parser::Fields {
                fields: HashMap::from([("field".to_string(), field)]),
            }),
            children: None,
        };
        let nodes = vec![node_a, node_b, node_c];
        let language = get_language(nodes);
        let state = State::new(&language, Config::default());
        let enum_tokens = state.get_enum(false);
        insta::assert_debug_snapshot!(snapshot_tokens(&enum_tokens));
    }
    #[test_log::test]
    fn test_add_field_subenums_missing_node() {
        let node_a = codegen_sdk_common::parser::Node {
            type_name: "node_a".to_string(),
            subtypes: vec![],
            named: false,
            root: false,
            fields: None,
            children: None,
        };
        let field = codegen_sdk_common::parser::FieldDefinition {
            types: vec![
                TypeDefinition {
                    type_name: "node_a".to_string(),
                    named: false,
                },
                TypeDefinition {
                    type_name: "node_b".to_string(),
                    named: true,
                },
            ],
            multiple: true,
            required: false,
        };
        let node_c = codegen_sdk_common::parser::Node {
            type_name: "node_c".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: Some(codegen_sdk_common::parser::Fields {
                fields: HashMap::from([("field".to_string(), field)]),
            }),
            children: None,
        };
        let nodes = vec![node_a, node_c];
        let language = get_language(nodes);
        let state = State::new(&language, Config::default());
        let enum_tokens = state.get_enum(false);
        let struct_tokens = state.get_structs();
        insta::assert_debug_snapshot!(snapshot_tokens(&enum_tokens));
        insta::assert_debug_snapshot!(snapshot_tokens(&struct_tokens));
    }
    #[test_log::test]
    fn test_get_structs() {
        let node = codegen_sdk_common::parser::Node {
            type_name: "test".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        };
        let nodes = vec![node];
        let language = get_language(nodes);
        let state = State::new(&language, Config::default());
        let struct_tokens = state.get_structs();
        insta::assert_debug_snapshot!(snapshot_tokens(&struct_tokens));
    }
    #[test_log::test]
    fn test_get_variants() {
        let node_a = codegen_sdk_common::parser::Node {
            type_name: "node_a".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        };
        let node_b = codegen_sdk_common::parser::Node {
            type_name: "node_b".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        };
        let parent = codegen_sdk_common::parser::Node {
            type_name: "parent".to_string(),
            subtypes: vec![
                TypeDefinition {
                    type_name: "node_a".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "node_b".to_string(),
                    named: true,
                },
            ],
            named: true,
            root: false,
            fields: None,
            children: None,
        };
        let nodes = vec![node_a, node_b, parent];
        let language = get_language(nodes);
        let state = State::new(&language, Config::default());
        let variants = state.get_variants("parent", true);
        assert_eq!(
            vec![
                TypeDefinition {
                    type_name: "node_a".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "node_b".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "comment".to_string(),
                    named: true,
                },
            ],
            variants
        );
        let variants = state.get_variants("parent", false);
        assert_eq!(
            vec![
                TypeDefinition {
                    type_name: "node_a".to_string(),
                    named: true,
                },
                TypeDefinition {
                    type_name: "node_b".to_string(),
                    named: true,
                },
            ],
            variants
        );
    }
    #[test_log::test]
    fn test_add_subenum() {
        let node_a = codegen_sdk_common::parser::Node {
            type_name: "node_a".to_string(),
            subtypes: vec![],
            named: true,
            root: false,
            fields: None,
            children: None,
        };
        let nodes = vec![node_a];
        let language = get_language(nodes);
        let mut state = State::new(&language, Config::default());

        state.add_subenum(
            "TestEnum",
            &vec![&TypeDefinition {
                type_name: "node_a".to_string(),
                named: true,
            }],
        );
        assert!(state.subenums.contains("TestEnum"));

        let node = state.nodes.get("NodeA").unwrap();
        assert!(node.subenums.contains(&"TestEnum".to_string()));
    }
}
