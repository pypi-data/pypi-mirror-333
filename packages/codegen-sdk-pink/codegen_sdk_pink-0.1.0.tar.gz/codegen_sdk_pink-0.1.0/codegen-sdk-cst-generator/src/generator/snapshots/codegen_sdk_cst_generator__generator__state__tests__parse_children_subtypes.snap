---
source: codegen-sdk-cst-generator/src/generator/state.rs
expression: snapshot_tokens(&enum_tokens)
---
#[subenum(AnonymousClassChildren, Definition)]
#[derive(Debug, Eq, PartialEq, Clone, Hash, salsa::Update, Delegate, Ord, PartialOrd)]
#[delegate(CSTNode<'db1>)]
pub enum NodeTypes<'db1> {
    AnonymousClass(AnonymousClass<'db1>),
    AnonymousFunction(AnonymousFunction<'db1>),
}
impl<'db3> From<AnonymousClass<'db3>> for NodeTypes<'db3> {
    fn from(node: AnonymousClass<'db3>) -> Self {
        Self::AnonymousClass(node)
    }
}
impl<'db3> From<AnonymousFunction<'db3>> for NodeTypes<'db3> {
    fn from(node: AnonymousFunction<'db3>) -> Self {
        Self::AnonymousFunction(node)
    }
}
impl<'db4> FromNode<'db4, NodeTypes<'db4>> for AnonymousClassChildren<'db4> {
    fn from_node(
        context: &mut ParseContext<'db4, NodeTypes<'db4>>,
        node: tree_sitter::Node,
    ) -> Result<(Self, Vec<indextree::NodeId>), ParseError> {
        match node.kind_id() {
            _ => {
                Err(ParseError::UnexpectedNode {
                    node_type: node.kind().to_string(),
                    backtrace: Backtrace::capture(),
                })
            }
        }
    }
}
impl<'db4> FromNode<'db4, NodeTypes<'db4>> for Definition<'db4> {
    fn from_node(
        context: &mut ParseContext<'db4, NodeTypes<'db4>>,
        node: tree_sitter::Node,
    ) -> Result<(Self, Vec<indextree::NodeId>), ParseError> {
        match node.kind_id() {
            _ => {
                Err(ParseError::UnexpectedNode {
                    node_type: node.kind().to_string(),
                    backtrace: Backtrace::capture(),
                })
            }
        }
    }
}
